src/xc_integrator/local_work_driver/device/scheme1_magma_base.cxx

```diff
@@ -33,9 +33,29 @@ void AoSScheme1MAGMABase::eval_xmat( double fac, XCDeviceData* _data, bool do_gr
   const auto submat_block_size = data->get_submat_chunk_size( nbf, 0 );
   auto static_stack  = data->static_stack;
   auto aos_stack     = data->aos_stack;
-  sym_pack_submat( ntasks, aos_stack.device_tasks, static_stack.dmat_device, 
+  double* dmat_ptr   = nullptr;
+  switch( den ) {
+    case DEN_S:
+      dmat_ptr = static_stack.dmat_s_device;
+      break;
+    case DEN_Z:
+      dmat_ptr = static_stack.dmat_z_device;
+      break;
+    case DEN_Y:
+      dmat_ptr = static_stack.dmat_y_device;
+      break;
+    case DEN_X:
+      dmat_ptr = static_stack.dmat_x_device;
+      break;
+    default:
+      GAUXC_GENERIC_EXCEPTION( "eval_xmat called with invalid density specifier" );
+  }
+  sym_pack_submat( ntasks, aos_stack.device_tasks, dmat_ptr,
     nbf, submat_block_size, data->device_backend_->queue() );
-
+  
+  // Update dmat on magma_stack if required
+  //if ( magma_stack.xdmat_array_device != dmat_ptr )
```

Comment:
If this is not being used elsewhere, just move the body here. No need to hide the details of a function (esp when there is data movement) if not needed